src/librustc/traits/select.rs

```diff
@@ -78,8 +77,40 @@ pub struct SelectionContext<'cx, 'tcx:'cx> {
     /// there is no type that the user could *actually name* that
     /// would satisfy it. This avoids crippling inference, basically.
     intercrate: bool,
+
+    /// Obligations generated during inference. If None, no obligations are to be collected (e.g.
+    /// this is swapped out to a `None` when probing).
+    inferred_obligations: Option<PredicateObligations<'tcx>>
```

Comment:
Hmm, so I don't know how I feel about this field being in here. I feel like in the case of `SelectionContext`, because it doesn't represent a single atomic operation, it'd be better to just thread the obligations through explicitly, rather than collecting them into a vector stored in the `selcx`.

The need to swap this to `None` when probing, for example, feels like a warning sign to me: probing shouldn't have to swap anything, we should be able to just throw away the obligations if we don't want them. 

OTOH, I can imagine that this is more efficient, since we'll probably create fewer "throw-away" vectors.